keras_rs/src/layers/retrieval/brute_force_retrieval.py

```diff
@@ -0,0 +1,172 @@
+from typing import Any, Optional, Text
+
+import keras
+
+from keras_rs.src import types
+from keras_rs.src.api_export import keras_rs_export
+
+
+@keras_rs_export("keras_rs.layers.BruteForceRetrieval")
+class BruteForceRetrieval(keras.layers.Layer):
+    """Brute force top-k retrieval.
+
+    This layer maintains a set of candidates and is able to exactly retrieve the
+    top-k candidates for a given query. It does this by computing the scores for
+    all of the candidates for the query and extracting the top ones. The
+    returned top-k candidates are sorted by score.
+
+    By default, this layer returns a tuple with the top scores and top
+    identifiers, but it can be configured to return a single tensor with the top
+    identifiers.
+
+    The identifiers for the candidates can be specified as a tensor. If not
+    provided, the IDs used are simply the candidate indices.
+    """
+
+    def __init__(
+        self,
+        candidate_embeddings: Optional[types.Tensor] = None,
```

Comment:
Do we actually need type annotations? They aren't in use in any other package. My take is that we should skip them for non-standard types (e.g. you can annotate int or string or bool but not Tensor).

Reply:
That's a good question. This is a bit of an experiment that I took from the internal prototype. My findings so far are that:

- I would be doable because Keras RS should be fairly simple and shouldn't have many cases of arguments that can be one of many different types.
- I personally like annotating tensors to make it clear when it's a single tensor vs. a tuple of tensors vs. an arbitrary structure.
- Type annotation is an all-or-nothing deal. It doesn't really work to partially annotate the code. One main reason is that you cannot do any basic verification (I'm using `mypy`) to verify that they're not completely wrong (which is worse than no annotation).
- The fact that Keras is not annotated (which would be really hard) diminishes the value of annotating here.

Bottom line, I can remove if you prefer.